// elf-tools/lib/constants.js
const constants = {};

constants.ELF_SIGNATURE_BYTES = Buffer.from('\x7fELF', 'ascii');

constants.class = {
  0: 'none',
  1: '32',
  2: '64',
  none: 0,
  '32': 1,
  '64': 2,
};

constants.endian = {
  0: 'none',
  1: 'lsb',
  2: 'msb',
  none: 0,
  lsb: 1,
  msb: 2,
};

constants.version = {
  0: 'none',
  1: 'current',
  none: 0,
  current: 1,
};

constants.osabi = {
  0: 'sysv',
  1: 'unix-sysv',
  2: 'netbsd',
  3: 'linux',
  6: 'solaris',
  7: 'aix',
  8: 'irix',
  9: 'freebsd',
  10: 'tru64',
  11: 'modesto',
  12: 'openbsd',
  13: 'openvms',
  14: 'nsk',
  15: 'aros',
  97: 'arm',
  255: 'standalone',
  sysv: 0,
  'unix-sysv': 1,
  netbsd: 2,
  linux: 3,
  solaris: 6,
  aix: 7,
  irix: 8,
  freebsd: 9,
  tru64: 10,
  modesto: 11,
  openbsd: 12,
  openvms: 13,
  nsk: 14,
  aros: 15,
  arm: 97,
  standalone: 255,
};

constants.abiversion = {
  0: 'none',
  1: 'current',
  none: 0,
  current: 1,
};

constants.machine = {
  0: 'none',
  1: 'm32',
  2: 'sparc',
  3: 'x86',
  4: '68k',
  5: '88k',
  6: '486',
  7: '860',
  8: 'mips',
  9: 's370',
  10: 'mips_rs3_le',
  11: 'rs6000',
  15: 'pa_risc',
  16: 'ncube',
  17: 'vpp500',
  18: 'sparc32plus',
  19: '960',
  20: 'ppc',
  21: 'ppc64',
  22: 's390',
  36: 'v800',
  37: 'fr20',
  38: 'rh32',
  39: 'rce',
  40: 'arm',
  41: 'alpha',
  42: 'sh',
  43: 'sparcv9',
  44: 'tricore',
  45: 'arc',
  46: 'h8_300',
  47: 'h8_300h',
  48: 'h8s',
  49: 'h8_500',
  50: 'ia_64',
  51: 'mips_x',
  52: 'coldfire',
  53: '68hc12',
  54: 'mma',
  55: 'pcp',
  56: 'ncpu',
  57: 'ndr1',
  58: 'starcore',
  59: 'me16',
  60: 'st100',
  61: 'tinyj',
  62: 'amd64',
  63: 'pdsp',
  66: 'fx66',
  67: 'st9plus',
  68: 'st7',
  69: '68hc16',
  70: '68hc11',
  71: '68hc08',
  72: '68hc05',
  73: 'svx',
  74: 'st19',
  75: 'vax',
  76: 'cris',
  77: 'javelin',
  78: 'firepath',
  79: 'zsp',
  80: 'mmix',
  81: 'huany',
  82: 'prism',
  83: 'avr',
  84: 'fr30',
  85: 'd10v',
  86: 'd30v',
  87: 'v850',
  88: 'm32r',
  89: 'mn10300',
  90: 'mn10200',
  91: 'pj',
  92: 'openrisc',
  93: 'arc_a5',
  94: 'xtensa',
  none: 0,
  m32: 1,
  sparc: 2,
  x86: 3,
  '68k': 4,
  '88k': 5,
  mips: 8,
  s370: 9,
  mips_rs3_le: 10,
  rs6000: 11,
  pa_risc: 15,
  ncube: 16,
  vpp500: 17,
  sparc32plus: 18,
  ppc: 20,
  ppc64: 21,
  s390: 22,
  v800: 36,
  fr20: 37,
  rh32: 38,
  rce: 39,
  arm: 40,
  alpha: 41,
  sh: 42,
  sparcv9: 43,
  tricore: 44,
  arc: 45,
  h8_300: 46,
  h8_300h: 47,
  h8s: 48,
  h8_500: 49,
  ia_64: 50,
  mips_x: 51,
  coldfire: 52,
  '68hc12': 53,
  mma: 54,
  pcp: 55,
  ncpu: 56,
  ndr1: 57,
  starcore: 58,
  me16: 59,
  st100: 60,
  tinyj: 61,
  amd64: 62,
  pdsp: 63,
  fx66: 66,
  st9plus: 67,
  st7: 68,
  '68hc16': 69,
  '68hc11': 70,
  '68hc08': 71,
  '68hc05': 72,
  svx: 73,
  st19: 74,
  vax: 75,
  cris: 76,
  javelin: 77,
  firepath: 78,
  zsp: 79,
  mmix: 80,
  huany: 81,
  prism: 82,
  avr: 83,
  fr30: 84,
  d10v: 85,
  d30v: 86,
  v850: 87,
  m32r: 88,
  mn10300: 89,
  mn10200: 90,
  pj: 91,
  openrisc: 92,
  arc_a5: 93,
  xtensa: 94,
};

constants.type = {
  '0': 'none',
  '1': 'rel',
  '2': 'exec',
  '3': 'dyn',
  '4': 'core',
  none: 0,
  rel: 1,
  exec: 2,
  dyn: 3,
  core: 4,
};

constants.entryType = {
  0: 'null',
  1: 'load',
  2: 'dynamic',
  3: 'interp',
  4: 'note',
  5: 'shlib',
  6: 'phdr',
  7: 'tls',
  null: 0,
  load: 1,
  dynamic: 2,
  interp: 3,
  note: 4,
  shlib: 5,
  phdr: 6,
  tls: 7,
  0x6464e550: 'sunw_unwind',
  0x6474e550: 'sunw_eh_frame',
  0x6474e551: 'gnu_stack',
  0x6474e552: 'gnu_relro',
  0x6ffffffa: 'losunw',
  0x6ffffffa: 'sunwbss',
  0x6ffffffb: 'sunwstack',
  0x6ffffffc: 'sunwdtrace',
  0x6ffffffd: 'sunwcap',
};

constants.entryFlags = {
  4: 'read',
  2: 'write',
  1: 'execute',
  read: 4,
  write: 2,
  execute: 1,
  0x00100000: 'sunw_failure',
  0x00200000: 'sunw_killed',
  0x00400000: 'sunw_siginfo',
};

constants.sectType = {
  0: 'null',
  1: 'progbits',
  2: 'symtab',
  3: 'strtab',
  4: 'rela',
  5: 'hash',
  6: 'dynamic',
  7: 'note',
  8: 'nobits',
  9: 'rel',
  10: 'shlib',
  11: 'dynsym',
  12: 'unknown12',
  13: 'unknown13',
  14: 'init_array',
  15: 'fini_array',
  16: 'preinit_array',
  17: 'group',
  18: 'symtab_shndx',

  null: 0,
  progbits: 1,
  symtab: 2,
  strtab: 3,
  rela: 4,
  hash: 5,
  dynamic: 6,
  note: 7,
  nobits: 8,
  rel: 9,
  shlib: 10,
  dynsym: 11,
  unknown12: 12,
  unknown13: 13,
  init_array: 14,
  fini_array: 15,
  preinit_array: 16,
  group: 17,
  symtab_shndx: 18,

  0x60000000: 'loos',
  0x6fffffef: 'sunw_capchain',
  0x6ffffff0: 'sunw_capinfo',
  0x6ffffff1: 'sunw_symsort',
  0x6ffffff2: 'sunw_tlssort',
  0x6ffffff3: 'sunw_ldynsym',
  0x6ffffff4: 'sunw_dof',
  0x6ffffff5: 'sunw_cap',
  0x6ffffff6: 'sunw_signature',
  0x6ffffff7: 'sunw_annotate',
  0x6ffffff8: 'sunw_debugstr',
  0x6ffffff9: 'sunw_debug',
  0x6ffffffa: 'sunw_move',
  0x6ffffffb: 'sunw_comdat',
  0x6ffffffc: 'sunw_syminfo',
  0x6ffffffd: 'sunw_verdef',
  0x6ffffffe: 'sunw_verneed',
  0x6fffffff: 'sunw_versym',
  0x6fffffff: 'hisunw',
  0x6fffffff: 'hios',
  0x6ffffff5: 'gnu_attributes',
  0x6ffffff6: 'gnu_hash',
  0x6ffffff7: 'gnu_liblist',
  0x6ffffff8: 'checksum',
};

constants.sectFlags = {
  0x01: 'write',
  0x02: 'alloc',
  0x04: 'execute',
  0x10: 'merge',
  0x20: 'strings',
  0x40: 'info_link',
  0x80: 'link_order',
  0x100: 'os_nonconforming',
  0x200: 'group',
  0x400: 'tls',
  write: 0x1,
  alloc: 0x2,
  execute: 0x4,
  merge: 0x10,
  strings: 0x20,
  info_link: 0x40,
  link_order: 0x80,
  os_nonconforming: 0x100,
  group: 0x200,
  tls: 0x400,
};

constants.dynTypes = {
  '0': 'DT_NULL',
  '1': 'DT_NEEDED',
  '2': 'DT_PLTRELSZ',
  '3': 'DT_PLTGOT',
  '4': 'DT_HASH',
  '5': 'DT_STRTAB',
  '6': 'DT_SYMTAB',
  '7': 'DT_RELA',
  '8': 'DT_RELASZ',
  '9': 'DT_RELAENT',
  '10': 'DT_STRSZ',
  '11': 'DT_SYMENT',
  '12': 'DT_INIT',
  '13': 'DT_FINI',
  '14': 'DT_SONAME',
  '15': 'DT_RPATH',
  '16': 'DT_SYMBOLIC',
  '17': 'DT_REL',
  '18': 'DT_RELSZ',
  '19': 'DT_RELENT',
  '20': 'DT_PLTREL',
  '21': 'DT_DEBUG',
  '22': 'DT_TEXTREL',
  '23': 'DT_JMPREL',
  '24': 'DT_BIND_NOW',
  '25': 'DT_INIT_ARRAY',
  '26': 'DT_FINI_ARRAY',
  '27': 'DT_INIT_ARRAYSZ',
  '28': 'DT_FINI_ARRAYSZ',
  '30': 'DT_FLAGS',
  0x6ffffef5: 'DT_GNU_HASH',
  0x6ffffffb: 'DT_FLAGS_1',
  0x6ffffffe: 'DT_VERNEED',
  0x6fffffff: 'DT_VERNEEDNUM',
  0x6ffffff0: 'DT_VERSYM',
  0x6ffffff9: 'DT_RELACOUNT',
  
  DT_NULL: 0,
  DT_NEEDED: 1,
  DT_PLTRELSZ: 2,
  DT_PLTGOT: 3,
  DT_HASH: 4,
  DT_STRTAB: 5,
  DT_SYMTAB: 6,
  DT_RELA: 7,
  DT_RELASZ: 8,
  DT_RELAENT: 9,
  DT_STRSZ: 10,
  DT_SYMENT: 11,
  DT_INIT: 12,
  DT_FINI: 13,
  DT_SONAME: 14,
  DT_RPATH: 15,
  DT_SYMBOLIC: 16,
  DT_REL: 17,
  DT_RELSZ: 18,
  DT_RELENT: 19,
  DT_PLTREL: 20,
  DT_DEBUG: 21,
  DT_TEXTREL: 22,
  DT_JMPREL: 23,
  DT_BIND_NOW: 24,
  DT_INIT_ARRAY: 25,
  DT_FINI_ARRAY: 26,
  DT_INIT_ARRAYSZ: 27,
  DT_FINI_ARRAYSZ: 28,
  DT_FLAGS: 30,
  DT_GNU_HASH: 0x6ffffef5,
  DT_FLAGS_1: 0x6ffffffb,
  DT_VERNEED: 0x6ffffffe,
  DT_VERNEEDNUM: 0x6fffffff,
  DT_VERSYM: 0x6ffffff0,
  DT_RELACOUNT: 0x6ffffff9,
};

module.exports = constants;
