// log-process-errors/build/src/level.js
"use strict";Object.defineProperty(exports,"__esModule",{value:true});exports.validateLevels=exports.getExampleLevels=exports.applyDefaultLevels=exports.getLevel=void 0;var _process=require("process");

var _filterObj=_interopRequireDefault(require("filter-obj"));
var _jestValidate=require("jest-validate");
var _mapObj=_interopRequireDefault(require("map-obj"));

var _utils=require("./utils.js");function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj}}


const getLevel=function({opts,name,error}){
const level=(0,_utils.result)(opts.level[name],error);

if(level==="default"||level===undefined){
return DEFAULT_LEVEL[name];
}

if(LEVELS.has(level)){
return level;
}

(0,_process.emitWarning)(
`Invalid option 'level.${name}' returning '${level}': function must return one of ${LEVELS_ARR.join(
", ")
}`);


return DEFAULT_LEVEL[name];
};exports.getLevel=getLevel;


const applyDefaultLevels=function({
opts:{level:{default:defaultLevel,...level}={}}})
{
const levelA=(0,_filterObj.default)(level,isDefined);

if(defaultLevel===undefined){
return{...DEFAULT_LEVEL,...levelA};
}

const defaultLevels=(0,_mapObj.default)(DEFAULT_LEVEL,eventName=>[
eventName,
defaultLevel]);

return{...defaultLevels,...levelA};
};exports.applyDefaultLevels=applyDefaultLevels;

const isDefined=function(key,value){
return value!==undefined;
};


const getExampleLevels=function(){
return(0,_mapObj.default)(DEFAULT_LEVEL,getExampleLevel);
};exports.getExampleLevels=getExampleLevels;

const getExampleLevel=function(eventName,level){

return[eventName,(0,_jestValidate.multipleValidOptions)(level,()=>{})];
};

const validateLevels=function({level}){
Object.entries(level).forEach(validateLevel);
};exports.validateLevels=validateLevels;

const validateLevel=function([name,level]){
if(isValidLevel({level})){
return;
}

throw new Error(
`Invalid option 'level.${name}' '${level}': must be a function or one of ${LEVELS_ARR.join(
", ")
}`);

};

const isValidLevel=function({level}){
return LEVELS.has(level)||level===undefined||typeof level==="function";
};

const LEVELS_ARR=["debug","info","warn","error","silent","default"];
const LEVELS=new Set(LEVELS_ARR);

const DEFAULT_LEVEL={
default:"error",
uncaughtException:"error",
warning:"warn",
unhandledRejection:"error",
rejectionHandled:"error",
multipleResolves:"info"};
